src/connectors/query-rules/connectQueryRules.ts

```diff
@@ -0,0 +1,237 @@
+import noop from 'lodash/noop';
+import isEqual from 'lodash/isEqual';
+import {
+  Renderer,
+  RenderOptions,
+  WidgetFactory,
+  Helper,
+  FacetRefinement,
+} from '../../types';
+import {
+  checkRendering,
+  createDocumentationMessageGenerator,
+  warning,
+} from '../../lib/utils';
+
+const withUsage = createDocumentationMessageGenerator({
+  name: 'query-rules',
+  connector: true,
+});
+
+type ParamTrackedFilters = {
+  [facetName: string]: (facetValues: string[]) => string[];
+};
+type ParamTransformRuleContexts = (ruleContexts: string[]) => string[];
+type ParamTransformItems = (items: object[]) => any;
+
+export type QueryRulesConnectorParams = {
+  trackedFilters?: ParamTrackedFilters;
+  transformRuleContexts?: ParamTransformRuleContexts;
+  transformItems?: ParamTransformItems;
+};
+
+export interface QueryRulesRenderOptions<T> extends RenderOptions<T> {
+  userData: object[];
+}
+
+export type QueryRulesRenderer<T> = Renderer<QueryRulesRenderOptions<T>>;
+
+export type QueryRulesWidgetFactory<T> = WidgetFactory<
+  QueryRulesConnectorParams & T
+>;
+
+export type QueryRulesConnector = {
+  <T>(
+    render: QueryRulesRenderer<T>,
+    unmount?: () => void
+  ): QueryRulesWidgetFactory<T>;
+};
+
+// A context rule must consist only of alphanumeric characters, hyphens, and underscores.
+// See https://www.algolia.com/doc/guides/managing-results/refine-results/merchandising-and-promoting/in-depth/implementing-query-rules/#context
+function escapeRuleContext(ruleName: string) {
+  return ruleName.replace(/[^a-z0-9-_]+/gi, '_');
+}
+
+function getRuleContextsFromTrackedFilters({
+  helper,
+  trackedFilters,
+}: {
+  helper: Helper;
+  trackedFilters: ParamTrackedFilters;
+}) {
+  const ruleContexts = Object.keys(trackedFilters).reduce(
+    (facets: string[], facetName) => {
```

Comment:
```suggestion
    (facets, facetName) => {
```

We can avoid it with the generic of `reduce<string[]>()`.

Reply:
Cool, thank you 👍